File: packages/stera_recorder/android/src/main/kotlin/open/fpvlabs/stera/ar_recorder/ArRecorderMethodChannelHandler.kt
package open.fpvlabs.stera.ar_recorder

import android.app.Activity
import android.os.Handler
import android.os.Looper
import android.view.WindowManager
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.view.TextureRegistry
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * Handles Flutter MethodChannel calls for AR recording operations.
 */
class ArRecorderMethodChannelHandler(
    private val activity: Activity,
    binaryMessenger: BinaryMessenger,
    private val textureRegistry: TextureRegistry
) {

    companion object {
        private const val CHANNEL_NAME = "ar_recorder"
    }

    private val arRecorder: ArRecorder = ArRecorderImpl(activity, textureRegistry)
    private val mainHandler = Handler(Looper.getMainLooper())
    private val controlExecutor: ExecutorService = Executors.newSingleThreadExecutor()

    private val methodChannel: MethodChannel = MethodChannel(binaryMessenger, CHANNEL_NAME)

    init {
        methodChannel.setMethodCallHandler { call, result ->
            when (call.method) {
                "initializeSession" -> handleInitializeSession(call, result)
                "startRecording" -> handleStartRecording(call, result)
                "pauseRecording" -> handlePauseRecording(result)
                "resumeRecording" -> handleResumeRecording(result)
                "cancelRecording" -> handleCancelRecording(result)
                "stopRecording" -> handleStopRecording(result)
                "getRecordingState" -> handleGetRecordingState(result)
                "checkArAvailability" -> handleCheckArAvailability(result)
                "disposeSession" -> handleDisposeSession(result)
                else -> result.notImplemented()
            }
        }
        println("✅ ArRecorderMethodChannelHandler initialized")
    }

    /**
     * Handles initializeSession method call.
     * Must run on main thread because textureRegistry.createSurfaceTexture() requires a Looper.
     */
    private fun handleInitializeSession(call: MethodCall, result: MethodChannel.Result) {
        try {
            activity.window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
            val response = arRecorder.initializeSession(call.arguments)
            result.success(response)
        } catch (e: Exception) {
            activity.window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
            e.printStackTrace()
            result.error(
                "INITIALIZATION_ERROR",
                e.message ?: "Failed to initialize AR session",
                null
            )
        }
    }

    /**
     * Handles startRecording method call.
     */
    private fun handleStartRecording(call: MethodCall, result: MethodChannel.Result) {
        runBackgroundAction(
            action = { arRecorder.startRecording(call.arguments) },
            errorCode = "START_RECORDING_ERROR",
            errorMessage = "Failed to start recording",
            result = result
        )
    }

    private fun handlePauseRecording(result: MethodChannel.Result) {
        runBackgroundAction(
            action = { arRecorder.pauseRecording() },
            errorCode = "PAUSE_RECORDING_ERROR",
            errorMessage = "Failed to pause recording",
            result = result
        )
    }

    private fun handleResumeRecording(result: MethodChannel.Result) {
        runBackgroundAction(
            action = { arRecorder.resumeRecording() },
            errorCode = "RESUME_RECORDING_ERROR",
            errorMessage = "Failed to resume recording",
            result = result
        )
    }

    private fun handleCancelRecording(result: MethodChannel.Result) {
        runBackgroundAction(
            action = { arRecorder.cancelRecording() },
            errorCode = "CANCEL_RECORDING_ERROR",
            errorMessage = "Failed to cancel recording",
            result = result
        )
    }

    /**
     * Handles stopRecording method call.
     */
    private fun handleStopRecording(result: MethodChannel.Result) {
        runBackgroundAction(
            action = { arRecorder.stopRecording() },
            errorCode = "STOP_RECORDING_ERROR",
            errorMessage = "Failed to stop recording",
            result = result
        )
    }

    private fun runBackgroundAction(
        action: () -> Map<String, Any?>,
        errorCode: String,
        errorMessage: String,
        result: MethodChannel.Result
    ) {
        controlExecutor.execute {
            try {
                val response = action()
                mainHandler.post {
                    result.success(response)
                }
            } catch (e: Exception) {
                e.printStackTrace()
                mainHandler.post {
                    result.error(
                        errorCode,
                        e.message ?: errorMessage,
                        null
                    )
                }
            }
        }
    }

    /**
     * Handles getRecordingState method call.
     */
    private fun handleGetRecordingState(result: MethodChannel.Result) {
        try {
            val response = arRecorder.getRecordingState()
            result.success(response)
        } catch (e: Exception) {
            e.printStackTrace()
            result.error(
                "STATE_ERROR",
                e.message ?: "Failed to get recording state",
                null
            )
        }
    }

    /**
     * Handles checkArAvailability method call.
     * Safe to call on main thread — uses the synchronous cached check.
     */
    private fun handleCheckArAvailability(result: MethodChannel.Result) {
        try {
            val response = arRecorder.checkArAvailability()
            result.success(response)
        } catch (e: Exception) {
            e.printStackTrace()
            result.error(
                "AR_AVAILABILITY_ERROR",
                e.message ?: "Failed to check AR availability",
                null
            )
        }
    }

    /**
     * Handles disposeSession method call.
     * Must run on main thread because textureRegistry.release() requires UI thread.
     */
    private fun handleDisposeSession(result: MethodChannel.Result) {
        try {
            arRecorder.disposeSession()
            activity.window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
            result.success(mapOf("success" to true))
        } catch (e: Exception) {
            activity.window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
            e.printStackTrace()
            result.error(
                "DISPOSE_ERROR",
                e.message ?: "Failed to dispose session",
                null
            )
        }
    }

    /**
     * Releases native resources when the hosting activity is torn down.
     * Safe to call multiple times.
     */
    fun dispose() {
        try {
            arRecorder.disposeSession()
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            controlExecutor.shutdown()
        }
    }
}

File: packages/stera_recorder/ios/stera_recorder/Sources/stera_recorder/ArRecorderMethodChannelHandler.swift
import Foundation
import AVFoundation
import Flutter
import UIKit

/// Handles Flutter MethodChannel calls for AR recording operations.
class ArRecorderMethodChannelHandler {

    private static let channelName = "ar_recorder"

    private let interop = ARRecorderInterop.sharedInterop()
    private let cueAudioRouter = RecorderCueAudioRouter()
    private let methodChannel: FlutterMethodChannel
    private let lifecycleQueue = DispatchQueue(label: "ar.recorder.lifecycle", qos: .userInitiated)

    init(binaryMessenger: FlutterBinaryMessenger, textureRegistry: FlutterTextureRegistry) {
        self.methodChannel = FlutterMethodChannel(name: Self.channelName, binaryMessenger: binaryMessenger)
        self.interop.configure(textureRegistry: textureRegistry)

        self.interop.onLowStorageWarning = { [weak self] in
            DispatchQueue.main.async {
                self?.methodChannel.invokeMethod("lowStorageWarning", arguments: nil)
            }
        }
        self.interop.onStorageCritical = { [weak self] in
            DispatchQueue.main.async {
                self?.methodChannel.invokeMethod("storageCritical", arguments: nil)
            }
        }

        methodChannel.setMethodCallHandler { [weak self] call, result in
            guard let self = self else { return }
            switch call.method {
            case "initializeSession":
                self.handleInitializeSession(arguments: call.arguments, result: result)
            case "startRecording":
                self.handleStartRecording(arguments: call.arguments, result: result)
            case "pauseRecording":
                self.handlePauseRecording(result: result)
            case "resumeRecording":
                self.handleResumeRecording(result: result)
            case "cancelRecording":
                self.handleCancelRecording(result: result)
            case "stopRecording":
                self.handleStopRecording(result: result)
            case "getRecordingState":
                self.handleGetRecordingState(result: result)
            case "checkArAvailability":
                self.handleCheckArAvailability(result: result)
            case "disposeSession":
                self.handleDisposeSession(result: result)
            default:
                result(FlutterMethodNotImplemented)
            }
        }
        print("ArRecorderMethodChannelHandler initialized")
    }

    private func handleInitializeSession(arguments: Any?, result: @escaping FlutterResult) {
        UIApplication.shared.isIdleTimerDisabled = true
        let prefs = arguments as? [String: Any?]
        let response = interop.initializeSessionResponse(preferences: prefs)
        result(response)
    }

    private func handleStartRecording(arguments: Any?, result: @escaping FlutterResult) {
        let config = ARFRecordingConfig.fromNativeDictionary(arguments as? [String: Any])
        runBackgroundAction(
            action: { [weak self] in self?.interop.startRecording(config: config).toDictionary() ?? ["success": false, "error": "Handler released"] },
            errorCode: "START_RECORDING_ERROR",
            errorMessage: "Failed to start recording",
            result: result
        )
    }

    private func handlePauseRecording(result: @escaping FlutterResult) {
        runBackgroundAction(
            action: { [weak self] in self?.interop.pauseRecording().toDictionary() ?? ["success": false, "error": "Handler released"] },
            errorCode: "PAUSE_RECORDING_ERROR",
            errorMessage: "Failed to pause recording",
            result: result
        )
    }

    private func handleResumeRecording(result: @escaping FlutterResult) {
        runBackgroundAction(
            action: { [weak self] in self?.interop.resumeRecording().toDictionary() ?? ["success": false, "error": "Handler released"] },
            errorCode: "RESUME_RECORDING_ERROR",
            errorMessage: "Failed to resume recording",
            result: result
        )
    }

    private func handleCancelRecording(result: @escaping FlutterResult) {
        runBackgroundAction(
            action: { [weak self] in self?.interop.cancelRecording().toDictionary() ?? ["success": false, "error": "Handler released"] },
            errorCode: "CANCEL_RECORDING_ERROR",
            errorMessage: "Failed to cancel recording",
            result: result
        )
    }

    private func handleStopRecording(result: @escaping FlutterResult) {
        runBackgroundAction(
            action: { [weak self] in self?.interop.stopRecording().toDictionary() ?? ["success": false, "error": "Handler released"] },
            errorCode: "STOP_RECORDING_ERROR",
            errorMessage: "Failed to stop recording",
            result: result
        )
    }

    private func handleGetRecordingState(result: @escaping FlutterResult) {
        let response = interop.getRecordingState().toDictionary()
        result(response)
    }

    private func handleCheckArAvailability(result: @escaping FlutterResult) {
        let response = interop.checkArAvailability().toDictionary()
        result(response)
    }

    private func handleDisposeSession(result: @escaping FlutterResult) {
        lifecycleQueue.async { [weak self] in
            self?.interop.disposeSession()
            DispatchQueue.main.async {
                UIApplication.shared.isIdleTimerDisabled = false
                result(["success": true])
            }
        }
    }

    private func runBackgroundAction(
        action: @escaping () -> [String: Any?],
        errorCode: String,
        errorMessage: String,
        result: @escaping FlutterResult
    ) {
        lifecycleQueue.async {
            let response = action()
            DispatchQueue.main.async {
                result(response)
            }
        }
    }
}

private final class RecorderCueAudioRouter {
    private let session = AVAudioSession.sharedInstance()
    private var storedCategory: AVAudioSession.Category?
    private var storedMode: AVAudioSession.Mode?
    private var storedOptions: AVAudioSession.CategoryOptions?
    private var hasStoredPreviousConfiguration = false

    func prepareSpeakerRoute() throws {
        if !hasStoredPreviousConfiguration {
            storedCategory = session.category
            storedMode = session.mode
            storedOptions = session.categoryOptions
            hasStoredPreviousConfiguration = true
        }

        try session.setCategory(.playAndRecord, mode: .default, options: [.defaultToSpeaker, .mixWithOthers])
        try session.setActive(true)
        try session.overrideOutputAudioPort(.speaker)
    }

    /// Ensures audio session is configured for recording input without
    /// overriding the output port.  Returns false when no input is available
    /// (e.g. the device has no microphone).
    func ensureRecordingSession() -> Bool {
        do {
            if !hasStoredPreviousConfiguration {
                storedCategory = session.category
                storedMode = session.mode
                storedOptions = session.categoryOptions
                hasStoredPreviousConfiguration = true
            }
            try session.setCategory(.playAndRecord, mode: .default, options: [.defaultToSpeaker, .mixWithOthers])
            try session.setActive(true)
        } catch {
            print("RecorderCueAudioRouter: ensureRecordingSession failed – \(error)")
            return false
        }
        return session.isInputAvailable
    }

    func releaseSpeakerRoute() throws {
        guard hasStoredPreviousConfiguration else { return }

        try? session.overrideOutputAudioPort(.none)
        if let storedCategory, let storedMode, let storedOptions {
            try session.setCategory(storedCategory, mode: storedMode, options: storedOptions)
            try session.setActive(true)
        }

        hasStoredPreviousConfiguration = false
        storedCategory = nil
        storedMode = nil
        storedOptions = nil
    }
}
